### DISCUSSION TITLE
Use ARB format as a resource format

### DISCUSSION THREAD

**Discussion by mosuem** [eemeli/message-resource-wg#10] [General]
Just a thought: Why not use the `ARB` format as specified [here](https://github.com/google/app-resource-bundle/wiki/ApplicationResourceBundleSpecification) for MessageFormat 2 messages? It is pretty agnostic to the message format itself, currently using "MessageFormat 1", can encode metadata, and allows namespacing of messages. As it is built on top of JSON, it is easily human- and machine readable.

**Comment by eemeli** [comment]
ARB is a format where embedding MF2 messages would be really simple, and it does support metadata rather well. However, being based on JSON does not necessarily make it easy for humans to create or edit. While we may presume that in most cases translators will work with MF2 message resources through some tooling, that isn't necessarily always the case, especially for smaller organisations and entities.

My main objection for ARB, though, would be its lack of multi-line string support. Especially more complex MF2 messages are much easier to read when each `let`, `match` and `when` statement can be on a separate line.

**Reply by mosuem** [reply]
Thanks! What about using the ARB spec in YAML instead of JSON, giving more readability and multi-line strings?

**Reply by eemeli** [reply]
While YAML is certainly a valid container format for MF2 messages and better for humans than JSON, it still has a number of challenging aspects that we're hoping to improve on with the work here. In no particular order:
- Value quoting requirements depend on its first character due to YAML supporting JSON-y `{}` flow collections. For instance `{hello}` needs quoting, while `let $x = {$y} {{$x}}` does not.
- On comments, the YAML spec [says this](https://yaml.org/spec/1.2.2/#3233-comments):
  > Comments are a presentation detail and must not have any effect on the serialization tree or representation graph. In particular, comments are not associated with a particular node.

  Both of those aspects are counter to what's desirable of MF2 resources.
- YAML as a format is on one hand complex, and on the other optimised more for humans rather than computers. This means that parsing it is slower than parsing something like JSON. For MF2 resources, we should be able to define a resource format that's sufficiently legible and powerful while keeping its parsing complexity low and speed high.
